pkg/registry/manifest.go

```diff
@@ -30,76 +30,93 @@ func (c *Client) Manifest(image Image, dbManifest Manifest) (Manifest, error) {
 	ctx, cancel := c.timeoutContext()
 	defer cancel()
 
-	if c.sysCtx.DockerAuthConfig == nil {
-		c.sysCtx.DockerAuthConfig = &types.DockerAuthConfig{}
-		// TODO: Seek credentials
-		//auth, err := config.GetCredentials(c.sysCtx, reference.Domain(ref.DockerReference()))
-		//if err != nil {
-		//	return nil, errors.Wrap(err, "Cannot get registry credentials")
-		//}
-		//*c.sysCtx.DockerAuthConfig = auth
-	}
-
-	imgRef, err := ParseReference(image.String())
+	rmRef, err := ParseReference(image.String())
 	if err != nil {
 		return Manifest{}, errors.Wrap(err, "Cannot parse reference")
 	}
 
-	var imgDigest digest.Digest
-	if c.opts.CompareDigest {
-		imgDigest, err = docker.GetDigest(ctx, c.sysCtx, imgRef)
-		if err != nil {
-			return Manifest{}, errors.Wrap(err, "Cannot get image digest from HEAD request")
-		}
+	// Retrieve remote digest through HEAD request
+	rmDigest, err := docker.GetDigest(ctx, c.sysCtx, rmRef)
+	if err != nil {
+		return Manifest{}, errors.Wrap(err, "Cannot get image digest from HEAD request")
+	}
 
-		if dbManifest.Digest != "" && dbManifest.Digest == imgDigest {
-			return dbManifest, nil
-		}
+	// Digest match, returns db manifest
+	if c.opts.CompareDigest && len(dbManifest.Digest) > 0 && dbManifest.Digest == rmDigest {
+		return dbManifest, nil
 	}
 
-	imgCloser, err := imgRef.NewImage(ctx, c.sysCtx)
+	rmCloser, err := rmRef.NewImage(ctx, c.sysCtx)
 	if err != nil {
 		return Manifest{}, errors.Wrap(err, "Cannot create image closer")
 	}
-	defer imgCloser.Close()
+	defer rmCloser.Close()
 
-	rawManifest, _, err := imgCloser.Manifest(ctx)
+	rmRawManifest, _, err := rmCloser.Manifest(ctx)
 	if err != nil {
 		return Manifest{}, errors.Wrap(err, "Cannot get raw manifest")
 	}
 
-	if !c.opts.CompareDigest {
-		imgDigest, err = manifest.Digest(rawManifest)
+	// For manifests list compare also digest matching the platform
+	rmManifestMimeType := manifest.GuessMIMEType(rmRawManifest)
+	if c.opts.CompareDigest && len(dbManifest.Raw) > 0 && dbManifest.isManifestList() && isManifestList(rmManifestMimeType) {
+		dbManifestList, err := manifest.ListFromBlob(dbManifest.Raw, dbManifest.MIMEType)
+		if err != nil {
+			return Manifest{}, errors.Wrap(err, "Cannot parse manifest list")
+		}
+
+		dbManifestPlatformDigest, err := dbManifestList.ChooseInstance(c.sysCtx)
+		if err != nil {
+			return Manifest{}, errors.Wrapf(err, "Error choosing image instance")
+		}
+
+		rmManifestList, err := manifest.ListFromBlob(dbManifest.Raw, dbManifest.MIMEType)
+		if err != nil {
+			return Manifest{}, errors.Wrap(err, "Cannot parse manifest list")
+		}
+
+		rmManifestPlatformDigest, err := rmManifestList.ChooseInstance(c.sysCtx)
 		if err != nil {
-			return Manifest{}, errors.Wrap(err, "Cannot get digest")
+			return Manifest{}, errors.Wrapf(err, "Error choosing image instance")
+		}
+
+		if dbManifestPlatformDigest == rmManifestPlatformDigest {
+			return dbManifest, nil
```

Comment:
Shouldn’t this update the database with the new manifest list so that the next check can just use a HEAD?

Reply:
Yes not so accurate but I think it's fine to keep the current one as the digest match the platform described by the SystemContext anyway.

Reply:
Without that database update, every future check of this image will cost two GETs to reach this digest comparison, until the per-arch digest changes and the database entry is fully refreshed.

OTOH if you think the case all of this is dealing with (manifest list changes but the per-arch manifest doesn’t) is very rare and not worth optimizing for further, that’s perfectly fair and your decision entirely.

Reply:
Oh I thought you were talking about the manifest matching the platform. Good point.